# Describing data with summary statistics
---

This worksheet is tied to the Pandas Getting Started Tutorials, picking out particular tutorials to link them into a theme here.

We will focus on describing data.  This is the least risky in terms of bias and inaccurate conclusions as it should focus just on what data is presented to us.

Each exercise will ask you to work through on tutorial on the Getting Started page, to try the code from the tutorial here and to try a second, similar action.

---

The practice data from the tutorials comes from a dataset on Titanic passengers.


### Exercise 1 - open the Titanic dataset
---

The Titanic dataset is stored at this URL:
https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv

Read the dataset into a pandas dataframe that you will call **titanic**.

**Test output**:  
The shape of the dataframe will be (891, 12)

In [5]:
import pandas as pd
titanic = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv")
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### Exercise 2 - get summary information about the dataframe
---

Read through the tutorials:  
[What kind of data does pandas handle?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html#)  
[How do I read and write tabular data?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/02_read_write.html)

Use panda functions to display the following:
1.  A technical summary of the data (info())
2.  A description of the numerical data (describe())
3.  Display the Series 'Age'

**Test output**:   
1.  The info should show that there are only 204 values in the Cabin series, out of 891 records.  
2.  The description should show 7 columns and a mean age of 29.699118
3.  The Age series should have values of type float64 and Length 891

In [306]:
import pandas as pd
url="https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)
summary = (titanic.info())
display (summary)
description = (titanic.describe())
display (description)

titanic["Age"]




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


None

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

### Exercise 3 - aggregating statistics
---

Read through the tutorial:  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

Use panda functions to display the following summary statistics from the titanic dataset:  

1.  The average (mean) age of passengers  
2.  The median age and fare  
3.  The mean fare
4.  The modal fare and gender

**Test output**:   
29.699118, Age 28.0000 Fare 14.4542, 32.2042079685746, Fare 8.05 Sex male 


In [307]:
import pandas as pd
url="https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)

mean_age = titanic["Age"].mean()
display (mean_age)
median_age_fare = titanic[["Age", "Fare"]].median()
display (median_age_fare)
mean_fare = titanic["Fare"].mean()
display (mean_fare)
mode_fare_sex = titanic[["Fare", "Sex"]].mode()
display (mode_fare_sex)



29.69911764705882

Age     28.0000
Fare    14.4542
dtype: float64

32.2042079685746

Fare   Sex
0  8.05  male

### Exercise 4 - displaying other statistics
---

Take a look at the list of methods available for giving summary statistics [here](https://pandas.pydata.org/docs/user_guide/basics.html#basics-stats) 

Use panda functions, and your existing knowledge, to display the following summary statistics from the titanic dataset:

1.  The total number of passengers on the titanic
2.  The age of the youngest passenger
3.  The most expensive ticket price
4.  The range of ticket prices
5.  The number of passenges with cabins
6.  The code for the port where the highest number of passengers embarked
7.  The most populous gender
8.  The standard deviation for age and fare

**Test output**:  
891, 0.42, 512.3292, 512.3292, 204, S, male, Age 14.526497 Fare 49.693429

In [308]:
import pandas as pd
url ="https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)

passenger_count = titanic["PassengerId"].count()
display (passenger_count) #1

min_age = titanic["Age"].min()
display (min_age) #2

max_fare = titanic["Fare"].max()
display (max_fare) #3

range_fare = titanic["Fare"].max()-titanic["Fare"].min()
display (range_fare) #4

#5

port = titanic["Embarked"].mode()#6
display (port)
pop_gender = titanic["Sex"].mode()#7
display(pop_gender)

std_age_fare = titanic[["Age", "Fare"]].std()
display (std_age_fare)#*


891

0.42

512.3292

512.3292

0    S
dtype: object

0    male
dtype: object

Age     14.526497
Fare    49.693429
dtype: float64

### Exercise 5 - aggregating statistics grouped by category
---

Refer again to the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)   
looking particularly at the section on Aggregating statistics grouped by category.

1.  What is the mean age for male versus female Titanic passengers?
2.  What is the mean ticket fare price for each of the sex and cabin class combinations?
3.  What is the mean ticket fare price for passengers who embarked at each port?
4.  Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?

**Test output**:  
1.  female 27.915709 male 30.726645
2.  
```
female  1         106.125798
            2          21.970121
            3          16.118810
male    1          67.226127
            2          19.741782
            3          12.661633
```
3.  
```
C    59.954144
Q    13.276030
S    27.079812
```
4. 

```
Survived  Pclass

0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
```






In [309]:
import pandas as pd
url="https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)

mean_age = titanic[["Sex", "Age"]].groupby("Sex").mean() #1.
display (mean_age)

mean_ticket = titanic.groupby(["Sex", "Pclass"])["Fare"].mean() #2
display (mean_ticket)

mean_port_fare = titanic[["Embarked", "Fare"]].groupby("Embarked").mean() #3
display (mean_port_fare)

class_survived = titanic.groupby(["Survived", "Pclass"])["PassengerId"].count() #4
display (class_survived)


Age
Sex              
female  27.915709
male    30.726645

Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64

Fare
Embarked           
C         59.954144
Q         13.276030
S         27.079812

Survived  Pclass
0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
Name: PassengerId, dtype: int64

### Exercise 6 - an aggregation of different statistics
---

Use the function titanic.agg() as shown in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

1.  Display

```
     {
         "Age": ["min", "max", "median", "skew"],
         "Fare": ["min", "max", "median", "mean"]
     }
```
2.  Display:  
min, max and mean for Age  
min, max and standard deviation for Fare  
count for Cabin

**Test output**:   
1.  	
```
                  Age	      Fare  
max	   80.000000	512.329200  
mean	  NaN	      32.204208  
median	28.000000	14.454200  
min	   0.420000	 0.000000  
skew	  0.389108	 NaN
```

2.   
```
	        Age	    Fare	   Cabin
count  NaN	    NaN	    204.0
max	80.000000  512.329200 NaN
mean   29.699118  NaN        NaN
min	0.420000   0.000000   NaN
std	NaN        49.693429  NaN
```




In [310]:
import pandas as pd
url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)


one = titanic.agg(
            {
                "Age": ["min", "max", "median", "skew"],
                "Fare": ["min", "max", "median", "mean"],
            }
        )

two = titanic.agg(
            {
                "Age": ["min", "max", "mean"],
                "Fare": ["min", "max", "std"],
                "Cabin": ["count"]
                
            }
        

)

display(one, two)

Age        Fare
min      0.420000    0.000000
max     80.000000  512.329200
median  28.000000   14.454200
skew     0.389108         NaN
mean          NaN   32.204208

Age        Fare  Cabin
min     0.420000    0.000000    NaN
max    80.000000  512.329200    NaN
mean   29.699118         NaN    NaN
std          NaN   49.693429    NaN
count        NaN         NaN  204.0

### Exercise 7 - count by category
---

Read the section Count number of records by category in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)

1. Display the number of passengers of each gender who had a ticket
2. Display the number of passengers who embarked at each port and had a ticket
3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic (*Hint:  try getting the PassengerIds with a count for survived or not.  Store this value in a new variable, which will contain a list/array.  The second item in this list will be the number who survived.  You can use this number and the count of PassengerIds to calculate the percentage*)

**Test output**:  
1.  female 314, male 577
2.  C 168, Q 77, S 644
3.  38.38383838383838



In [311]:
import pandas as pd
url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)

gender_count = titanic.groupby("Sex")["Ticket"].count() #1

port_count = titanic.groupby("Embarked")["Ticket"].count() #2

survived_titanic = titanic[["Survived","PassengerId"]].groupby("Survived").count()
survived_list = survived_titanic.values.tolist()

total_passengers = titanic["PassengerId"].count()

percentage = (survived_list[1]/total_passengers)*100
display (gender_count, port_count, percentage)



Sex
female    314
male      577
Name: Ticket, dtype: int64

Embarked
C    168
Q     77
S    644
Name: Ticket, dtype: int64

array([38.38383838])

### Exercise 8 - summary happiness statistics
---

Open the data set here: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true

It contains data on people's perception of happiness levels in a number of countries across the world.

1.  Display the number of records in the set  
2.  Display the description of the numerical data  
3.  Display the highest GDP and life expectancy  
4.  Display the mean, max and min for Freedom,  mean, max, min and skew for Generosity and mean, min, max and std for GDP  

**Test output**:  
1.  156
2.  Table showing count, mean, std, min, 25%, 50%, 75%, max for 8 columns
3.  GDP 0.905147, life expectancy 0.725244  
4.  


```
	   Freedom to make life choices	Generosity	GDP per capita
max	 0.631000	                   0.566000	  1.684000
mean	0.392571	                   0.184846	  0.905147
min	 0.000000	                   0.000000  	0.000000
skew	NaN	                        0.745942	  NaN
std	 NaN	                        NaN	       0.398389
```




In [313]:
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true"
happiness_level = pd.read_excel(url)
print (len(happiness_level))

description = happiness_level.describe()

max_gdp_expectancy = happiness_level[["GDP per capita","Healthy life expectancy"]].max()
print (max_gdp_expectancy)

happiness = happiness_level.agg( {
                    "Freedom to make life choices" : ["mean", "max", "min"],
                    "Generosity" : ["mean", "max", "min", "skew"],
                    "GDP per capita" : ["mean", "max", "min", "std"]
}
)

display (description, max_gdp_expectancy, happiness)




156
GDP per capita             1.684
Healthy life expectancy    1.141
dtype: float64


Overall rank       Score  GDP per capita  Social support  \
count    156.000000  156.000000      156.000000      156.000000   
mean      78.500000    5.407096        0.905147        1.208814   
std       45.177428    1.113120        0.398389        0.299191   
min        1.000000    2.853000        0.000000        0.000000   
25%       39.750000    4.544500        0.602750        1.055750   
50%       78.500000    5.379500        0.960000        1.271500   
75%      117.250000    6.184500        1.232500        1.452500   
max      156.000000    7.769000        1.684000        1.624000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               156.000000                    156.000000  156.000000   
mean                  0.725244                      0.392571    0.184846   
std                   0.242124                      0.143289    0.095254   
min                   0.000000                      0.000000    0.000000   
25%                   0.547750                      0.308000    0.108750   
50%                   0.789000                      0.417000    0.177500   
75%                   0.881750                      0.507250    0.248250   
max                   1.141000                      0.631000    0.566000   

       Perceptions of corruption  
count                 156.000000  
mean                    0.110603  
std                     0.094538  
min                     0.000000  
25%                     0.047000  
50%                     0.085500  
75%                     0.141250  
max                     0.453000

GDP per capita             1.684
Healthy life expectancy    1.141
dtype: float64

Freedom to make life choices  Generosity  GDP per capita
mean                      0.392571    0.184846        0.905147
max                       0.631000    0.566000        1.684000
min                       0.000000    0.000000        0.000000
skew                           NaN    0.745942             NaN
std                            NaN         NaN        0.398389

### Exercise 9 - migration data
---

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true  Open the sheet named *Country Migration*

1.  Describe the dataset  
2.  Show summary information
3.  Display the mean net per 10K migration in each of the years 2015 to 2019
4.  Display the mean, max and min migration for the year 2019 for each of the regions (*base_country_wb_region*)
5.  Display the median net migration for the years 2015 and 2019 for the base countries by income level
6.  Display the number of target countries in each income level
7.  Display the mean net migration for all five years, for each income level

**Test output**:  
1  count, mean, std, min, 25%, 50%, 75%, max for 9 columns

2  shows 16 columns with non-null count of 4148 in each column

3  
```
net_per_10K_2015    0.461757
net_per_10K_2016    0.150248
net_per_10K_2017   -0.080272
net_per_10K_2018   -0.040591
net_per_10K_2019   -0.022743
dtype: float64
```

4  
```
	net_per_10K_2019
mean	max	min
base_country_wb_region			
East Asia & Pacific	0.198827	21.57	-9.88
Europe & Central Asia	0.208974	87.71	-21.34
Latin America & Caribbean	-0.904602	21.15	-31.75
Middle East & North Africa	-0.107655	55.60	-50.33
North America	0.239246	23.20	-0.29
South Asia	-0.514577	13.72	-24.89
Sub-Saharan Africa	-0.279729	37.11	-21.54
```

5  
```
	net_per_10K_2015	net_per_10K_2019
base_country_wb_income		
High Income	0.02	0.04
Low Income	0.42	-0.05
Lower Middle Income	-0.02	-0.07
Upper Middle Income	-0.03	-0.08
```

6  
```
base_country_wb_income
High Income            2415
Low Income              185
Lower Middle Income     653
Upper Middle Income     895
Name: target_country_name, dtype: int64 
```

7  
```
net_per_10K_2015	net_per_10K_2016	net_per_10K_2017	net_per_10K_2018	net_per_10K_2019
base_country_wb_income					
High Income	0.505482	0.391379	0.314178	0.379201	0.401470
Low Income	1.876432	0.798270	-0.684865	-0.677784	-0.681459
Lower Middle Income	0.591654	-0.029893	-0.519433	-0.527136	-0.476616
Upper Middle Income	-0.043419	-0.502916	-0.699240	-0.686626	-0.700101
```






In [314]:
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true"
country_migration = pd.read_excel(url, sheet_name="Country Migration")
description = country_migration.describe() #1
summary = country_migration.info() #2

#3 Display the mean net per 10K migration in each of the years 2015 to 2019
three = pd.DataFrame(country_migration,columns=['net_per_10K_2015','net_per_10K_2016','net_per_10K_2017','net_per_10K_2018','net_per_10K_2019'])
three1=(three.mean())

#4 Display the mean, max and min migration for the year 2019 for each of the regions (base_country_wb_region)
four = country_migration.groupby("base_country_wb_region").agg( {"net_per_10K_2019" : ["mean", "max","min" ]})
print (four)

#5 Display the median net migration for the years 2015 and 2019 for the base countries by income level
five = country_migration.groupby("base_country_wb_income").agg ({"net_per_10K_2015" : ["median"], "net_per_10K_2019":["median"]})
print (five)

#6 Display the number of target countries in each income level
six = country_migration.groupby("target_country_wb_income")["target_country_name"].count()
print (six)

#7 Display the mean net migration for all five years, for each income level
cols = {"net_per_10K_2019" : ["mean"],"net_per_10K_2018" : ["mean"],"net_per_10K_2017" : ["mean"]
        ,"net_per_10K_2016" : ["mean"],"net_per_10K_2015" : ["mean"]}
seven = country_migration.groupby("base_country_wb_income").agg(cols)
print(seven)

display (description, summary, three1, four, five, six, seven)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4148 entries, 0 to 4147
Data columns (total 26 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   base_country_code         4148 non-null   object 
 1   base_country_name         4148 non-null   object 
 2   base_lat                  4148 non-null   float64
 3   base_long                 4148 non-null   float64
 4   base_country_wb_income    4148 non-null   object 
 5   base_country_wb_region    4148 non-null   object 
 6   target_country_code       4148 non-null   object 
 7   target_country_name       4148 non-null   object 
 8   target_lat                4148 non-null   float64
 9   target_long               4148 non-null   float64
 10  target_country_wb_income  4148 non-null   object 
 11  target_country_wb_region  4148 non-null   object 
 12  net_per_10K_2015          4148 non-null   float64
 13  net_per_10K_2016          4148 non-null   float64
 14  net_per_

base_lat    base_long   target_lat  target_long  net_per_10K_2015  \
count  4148.000000  4148.000000  4148.000000  4148.000000       4148.000000   
mean     28.418022    21.698305    28.418022    21.698305          0.461757   
std      25.086012    61.937381    25.086012    61.937381          5.006530   
min     -40.900557  -106.346771   -40.900557  -106.346771        -37.010000   
25%      14.058324    -3.435973    14.058324    -3.435973         -0.150000   
50%      35.861660    19.145136    35.861660    19.145136          0.000000   
75%      47.516231    53.688046    47.516231    53.688046          0.240000   
max      64.963051   179.414413    64.963051   179.414413        150.680000   

       net_per_10K_2016  net_per_10K_2017  net_per_10K_2018  net_per_10K_2019  \
count       4148.000000       4148.000000       4148.000000       4148.000000   
mean           0.150248         -0.080272         -0.040591         -0.022743   
std            4.201118          3.203092          3.593876          3.633247   
min          -40.890000        -43.660000        -56.220000        -50.330000   
25%           -0.190000         -0.210000         -0.210000         -0.210000   
50%            0.000000          0.000000          0.000000          0.000000   
75%            0.220000          0.160000          0.170000          0.180000   
max          124.480000         87.000000         91.410000         87.710000   

       Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  \
count          0.0          0.0          0.0          0.0          0.0   
mean           NaN          NaN          NaN          NaN          NaN   
std            NaN          NaN          NaN          NaN          NaN   
min            NaN          NaN          NaN          NaN          NaN   
25%            NaN          NaN          NaN          NaN          NaN   
50%            NaN          NaN          NaN          NaN          NaN   
75%            NaN          NaN          NaN          NaN          NaN   
max            NaN          NaN          NaN          NaN          NaN   

       Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
count          0.0          0.0          0.0          0.0  
mean           NaN          NaN          NaN          NaN  
std            NaN          NaN          NaN          NaN  
min            NaN          NaN          NaN          NaN  
25%            NaN          NaN          NaN          NaN  
50%            NaN          NaN          NaN          NaN  
75%            NaN          NaN          NaN          NaN  
max            NaN          NaN          NaN          NaN

None

net_per_10K_2015    0.461757
net_per_10K_2016    0.150248
net_per_10K_2017   -0.080272
net_per_10K_2018   -0.040591
net_per_10K_2019   -0.022743
dtype: float64

net_per_10K_2019              
                                       mean    max    min
base_country_wb_region                                   
East Asia & Pacific                0.198827  21.57  -9.88
Europe & Central Asia              0.208974  87.71 -21.34
Latin America & Caribbean         -0.904602  21.15 -31.75
Middle East & North Africa        -0.107655  55.60 -50.33
North America                      0.239246  23.20  -0.29
South Asia                        -0.514577  13.72 -24.89
Sub-Saharan Africa                -0.279729  37.11 -21.54

net_per_10K_2015 net_per_10K_2019
                                 median           median
base_country_wb_income                                  
High Income                        0.02             0.04
Low Income                         0.42            -0.05
Lower Middle Income               -0.02            -0.07
Upper Middle Income               -0.03            -0.08

target_country_wb_income
High Income            2415
Low Income              185
Lower Middle Income     653
Upper Middle Income     895
Name: target_country_name, dtype: int64

net_per_10K_2019 net_per_10K_2018 net_per_10K_2017  \
                                   mean             mean             mean   
base_country_wb_income                                                      
High Income                    0.401470         0.379201         0.314178   
Low Income                    -0.681459        -0.677784        -0.684865   
Lower Middle Income           -0.476616        -0.527136        -0.519433   
Upper Middle Income           -0.700101        -0.686626        -0.699240   

                       net_per_10K_2016 net_per_10K_2015  
                                   mean             mean  
base_country_wb_income                                    
High Income                    0.391379         0.505482  
Low Income                     0.798270         1.876432  
Lower Middle Income           -0.029893         0.591654  
Upper Middle Income           -0.502916        -0.043419

### Exercise 10 - calculating range over a grouped series

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Open the sheet named *Skill Migration*

1.  Display the max for each skill group category of net migration for the year 2017
2.  Assign the max for each skill group category of net migration for the year 2017 to a variable called **max_skill_migration** and print `max_skill_migration`
3.  Create a second variable called **min_skill_migration** and assign to it the min for each skill group category of net migration for the year 2017, print `min_skill_migration`

4.  You now have two series `max_skill_migration` and `min_skill_migration` each of which is a numpy array.  You can perfom calculations on these two series in the same way as you would individual data items.

So, you can calculate the range for each skill category by subtracting the `min_skill_migration` from `max_skill_migration` to get a new series **skill_migration_range**

skill_migration_range = max_skill_migration - min_skill_migration

Try it out.

5.  Now calculate the range for the year 2019
6.  Now calculate the range for countries grouped by base country income level for the year 2015

**Test output**:  
1 and 2  
```
skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64
```

3    
```
skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64
```

4  
```
skill_group_category
Business Skills                4519.55
Disruptive Tech Skills         4124.75
Soft Skills                    4114.58
Specialized Industry Skills    8510.81
Tech Skills                    7397.76
Name: net_per_10K_2017, dtype: float64
```

5  
```
skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64
```

6  
```
wb_income
High income            4246.50
Low income             4556.42
Lower middle income    2148.36
Upper middle income    4045.43
Name: net_per_10K_2015, dtype: float64
```





In [315]:
def max_range_calc(a,b):

    skill_migration = pd.read_excel("https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true",sheet_name="Skill Migration")
    one = skill_migration.groupby(a)[b].max() #1
    display (one) #2
    two = skill_migration.groupby(a)[b].min()#3
    display (two)
    skill_migration_range = one - two
    display (skill_migration_range)#4

max_range_calc("skill_group_category","net_per_10K_2017")

def range_calc(a,b):
    skill_migration = pd.read_excel("https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true",sheet_name="Skill Migration")
    max = skill_migration.groupby(a)[b].max()
    min = skill_migration.groupby(a)[b].min()
    range = max - min
    display (range)

range_calc("skill_group_category","net_per_10K_2019") #5

range_calc("wb_income","net_per_10K_2015") #6
      

skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64

skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64

skill_group_category
Business Skills                4519.55
Disruptive Tech Skills         4124.75
Soft Skills                    4114.58
Specialized Industry Skills    8510.81
Tech Skills                    7397.76
Name: net_per_10K_2017, dtype: float64

skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64

wb_income
High income            4246.50
Low income             4556.42
Lower middle income    2148.36
Upper middle income    4045.43
Name: net_per_10K_2015, dtype: float64

# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer: 

## What caused you the most difficulty?

Your answer: 